### ISSUE TITLE
FR: Support `.ignore`/`.jjignore`, a Git Independent way to ignore files. 

### ISSUE THREAD

**Issue by PhilipMetzger** [martinvonz/jj#3525]
**Is your feature request related to a problem? Please describe.**

At some point we won't have Git as a base system anymore, which means we'll some way to migrate to a new `.ignore` file. It either could be a custom `.jjignore` or `.ignore` if it doesn't conflict with other tools you're using. This is to progress in 

**Describe the solution you'd like**
Add support for a `.jjignore`/`.ignore` file which can be adjacent to the `.gitignore` and a way to migrate the `.gitignore` to it (should be hopefully just a name change).  

**Describe alternatives you've considered**
Support `.gitignore` forever, which also could work, but I'd really like Git to be gone.

**Comment by thoughtpolice** [comment]
> While adjacent, the `jj` one would have `.git` in it while the `git` one would have `.jj` in it.

Note that you don't need to add `.jj` to a `.gitignore` today because when jj uses a colocated repo, it automatically adds `.jj/*` to the [special `.git/info/exclude`](https://luisdalmolin.dev/blog/ignoring-files-in-git-without-gitignore/) file.

**Comment by necauqua** [comment]
> A plain `.ignore` is not as future proof.

I kind of.. did not understand your message here. Why?

I've mentioned previously that creator of the [`ignore`](https://github.com/BurntSushi/ripgrep/tree/master/crates/ignore) crate (and ripgrep) is pushing for a unified .ignore (which is just a .gitignore that's not specific to git) file that's recognized by various tools and I like the idea a lot. So I'd be in favor of `.ignore` file compliant with that instead of making yet another `.somethingignore` file.

**Comment by joyously** [comment]
> I've mentioned previously that creator of the [`ignore`](https://github.com/BurntSushi/ripgrep/tree/master/crates/ignore) crate (and ripgrep) is pushing for a unified .ignore

I haven't seen that remark. I suppose it doesn't matter that all the VCSs would be listed in one file. That's what would be intended anyway. I haven't thought about it as much as the crate author, but when I wrote that about future proofing, I was thinking in terms of a future maintainer that doesn't really know the old tool so wouldn't know that generic `.ignore` goes with it, and also mixing ignore files from VCS and some other kind of tool like backups or encryption or something else not written yet.

**Comment by dbarnett** [comment]
I see two possible approaches here:
1. jj only respects .ignore in its own operations
2. jj attempts to bridge the gap with git and other backends to help them ignore the same files it's ignoring

Option 1 is easy but seems like it'll just tend to infuriate users in cases like colocated repos, where different VCS tools have to be taught separately to ignore files. Option 2 seems nicer but really finicky to get right.

**Comment by PhilipMetzger** [comment]
I think the second approach is the right one, as it allows a seamless transition from the Git backend to the future `native` backend, or any custom one if there's one downstream.

**Comment by dbarnett** [comment]
SGTM, but does that imply that a .ignore file in jj's store would get reflected with a _different_ filename in a git workspace? And would colocated repos end up with .ignore and .gitignore side-by-side?

Also FYI might require some changes in the ignore crate to support more involved inspection/reflection of the ignore specs.

**Comment by PhilipMetzger** [comment]
> SGTM, but does that imply that a .ignore file in jj's store would get reflected with a _different_ filename in a git workspace?

I don't think so, as for Git it just is a auxiliary file, in the worst case it could get added to the `.gitignore` itself, like how we exclude the `.jj` in colocated repositories. 

> And would colocated repos end up with .ignore and .gitignore side-by-side?

Probably, that means there should a repo level config knob to prioritize one of them. Although we should give our `.ignore` priority by default.

**Comment by joyously** [comment]
> > > I think we should do away with the ambiguous states of tracked/ignored and untracked/not ignored.

> > it could be that you keep ignores as is, but make the `track` and `untrack` command update the ignore file.

> could you expand on it?

The docs were stating that with a new config to disable auto-track, you have to edit the ignore file after using `track` or `untrack`. What are these commands doing if the ignore file is unchanged, and is it different with or without auto-track?
Are there really valuable states with the combination of "tracked", "untracked", and "ignored", or is it just one variable with two states: ignored/tracked ?
There is a race condition when you have to store the state in two places.
It seems to me that the command should bring up the ignore file in the editor, so the user can edit it. 
I think you have to define what "ignore" means first (see below). My concern with the disabling of auto-track is the file deletion on undo. If a file is ignored, it should never be touched. (What if the file in question is the `.gitignore`?)

Here are some links to Git and Breezy docs for their ignore handling:
**Git** has [`add`](https://git-scm.com/docs/git-add) and the index, [`clean`](https://git-scm.com/docs/git-clean) removes untracked files, [`gc`](https://git-scm.com/docs/git-gc) can affect files, [`rm`](https://git-scm.com/docs/git-rm) removes files, and [`config`](https://git-scm.com/docs/git-config) to set options. In the [doc for gitignore](https://git-scm.com/docs/gitignore), it says
> A gitignore file specifies intentionally untracked files that Git should ignore. Files already tracked by Git are not affected; see the NOTES

**Breezy** has [`add`](https://breezy-vcs.org/doc/en/user-reference/add-help.html) (no index), [`clean-tree`](https://breezy-vcs.org/doc/en/user-reference/clean-tree-help.html) removes unknown files, [`ignore` ](https://breezy-vcs.org/doc/en/user-reference/ignore-help.html)to update ignore list (creates `.bzrignore` if doesn't exist), `ignored` to list ignored files, [`remove`](https://breezy-vcs.org/doc/en/user-reference/remove-help.html) to remove files, and [`config` ](https://breezy-vcs.org/doc/en/user-reference/config-help.html)to set options. In the [doc for ignoring](https://breezy-vcs.org/doc/en/user-guide/controlling_registration.html), it says
> Note that ignore patterns are only matched against non-versioned files, and control whether they are treated as “unknown” or “ignored”. If a file is explicitly added, it remains versioned regardless of whether it matches an ignore pattern.

I don't know **Mercurial**, but here's a page on file handling, with [Permanently ignoring unwanted files](https://book.mercurial-scm.org/read/files.html#permanently-ignoring-unwanted-files-and-directories) near the bottom.

**Comment by PhilipMetzger** [comment]
I thought a bit more about unifying `untracked`/`ignored` file states the last two days and think that the only option for it is in a VFS, where in each workspace only has _the current files available_. 

> What are these commands doing if the ignore file is unchanged, and is it different with or without auto-track?

The only semantic difference for `jj` is that `track` will allow the snapshotting of otherwise ignored files. Ideally _there_ should be no difference between `untracked/ignored` but future and current users will need to adapt to it. 

> Are there really valuable states with the combination of "tracked", "untracked", and "ignored", or is it just one variable with two states: ignored/tracked ?

I think this assessment is correct, that it only should be one state and the third state is here for broader adoption. I'd still like to remove it at some point when the world gets used to having a clean repository.

**Comment by PhilipMetzger** [comment]
I've been thinking for a while now that if we go down the way of building our own `.jjignore`/`.jj-ignore` file, that it'd be extremely nice to support filesets in it, since they're more readable than RegEx patterns. IMO, that would be a nice to have.

**Comment by xpe** [comment]
> I've been thinking for a while now that if we go down the way of building our own `.jjignore`/`.jj-ignore` file, that it'd be extremely nice to support [filesets](https://docs.jj-vcs.dev/latest/filesets/) in it, since they're more readable than RegEx patterns. IMO, that would be a nice to have.

Yep! This is a good opportunity to create an improved file format. Worth designing well.

**Comment by ian-h-chamberlain** [comment]
> > I've been thinking for a while now that if we go down the way of building our own `.jjignore`/`.jj-ignore` file, that it'd be extremely nice to support [filesets](https://docs.jj-vcs.dev/latest/filesets/) in it, since they're more readable than RegEx patterns. IMO, that would be a nice to have.
> 
> Yep! This is a good opportunity to create an improved file format. Worth designing well.

Just a thought based on this idea, since there is some precedent: What about simply adding this as a new config field that can be adjusted via `jj config`? This would also allow for the existing config cascading:
- User-level excludes (like `~/.config/git/ignore`)
- Repo-level excludes (I guess this is closest to `.git/info/exclude`?)
- Workspace-level excludes (not sure Git has something like this)
- (still a missing piece) — something that could be checked in like a `.jj-config.toml`.

	This I'm less sure of, but it could be added separately as a follow-up using the same semantics. Making it more general-purpose than just an ignore file would be a different direction than some other proposals here, but also perhaps useful for devs who want to have some standard JJ configuration across their teams?

The parallel here is the existing `[revsets]` — maybe something like this?

```toml
[filesets]
ignore = 'default_ignores() | glob:"./some/other/path/*"'
# or maybe 
ignore = 'gitignore() | user.ignore() ~ "my-special-file"' # similar to .gitignore's !-prefix
```

Being able to explicitly reference the defaults might also be useful in defining how the cascading works. Each config file can simply override the "parent" level and choose how to merge the filesets, rather than prescribing that all cases are `|`'d together. (This is a pet peeve of mine with how `!` interacts between `.gitignore` and `.git/info/exclude`).

As some others in this thread have mentioned, the downside is that it would be JJ-specific and probably not easily shareable to other VCS tooling. There's also potentially confusing interactions with `snapshot.auto-track`, but I think that probably is going to be the case for any implementation of this kind of feature.

**Comment by josephlou5** [comment]
> * (still a missing piece) — something that could be checked in like a `.jj-config.toml`.
>   This I'm less sure of, but it could be added separately as a follow-up using the same semantics. Making it more general-purpose than just an ignore file would be a different direction than some other proposals here, but also perhaps useful for devs who want to have some standard JJ configuration across their teams?

See #8917 and its design doc https://docs.jj-vcs.dev/latest/design/managed-config/.

> The parallel here is the existing `[revsets]` — maybe something like this?
> ```
> [filesets]
> ...
> ```

I don't think it should be in a `[filesets]` table; this is an "ignore" config, and "filesets" are already a concept in jj for providing "a set of files" to commands. A `filesets` table could be confused as "default files for commands" (the way that `templates.*` and `revsets.*` are settings for commands) or could be confused as "fileset aliases" (https://docs.jj-vcs.dev/latest/filesets/#aliases). Certainly the "ignore" config should support / use filesets, but I think `filesets` is the wrong table name for it.
